### DISCUSSION TITLE
zod union vs XOR / exclusive OR

### DISCUSSION THREAD

**Discussion by jonathanstanley** [colinhacks/zod#2713] [General]
`z.union([...])` and `mySchema.or(...)` is NOT an XOR (exclusive or): https://github.com/colinhacks/zod/blob/f59be093ec21430d9f32bbcb628d7e39116adf34/src/types.ts#L2757-L2760

zod marks a success as soon as any successful schema is found. this is faster than XOR. 

In special cases, this might give surprising  results (ie: those receiving messy objects/unions from third parties without a discriminated union). FWIW: [XOR is also tricky in typescript](https://stackoverflow.com/questions/44425344/typescript-interface-with-xor-barstring-xor-cannumber).

see also
https://github.com/colinhacks/zod/issues/2325 union in objects error / undefined
https://github.com/colinhacks/zod/issues/2106 deprecating discriminated union?

**Comment by rubengmurray** [comment]
Was trying to find my own post (https://github.com/colinhacks/zod/discussions/2797) and came across this. Linking for reference.
